### PR TITLE
Testing tx/fee limits and max size

### PR DESCRIPTION
**Pull request by piotr-iohk**
# Issue Number

#469 
#511

# Overview

- [ ] Testing max inputs/outputs and max tx size for jormungandr and http-bridge

# Comments

Base branch is `piotr\estimate-testing` (I didn't want to pollute #511)

### CODE REVIEW

File: lib/core/test/integration/Test/Integration/Framework/DSL.hs [hunk 1]
```diff
@@ -613,6 +615,65 @@ emptyWalletWith ctx (name, passphrase, addrPoolGap) = do
...
+    wDest3 <- emptyWalletWith ctx ("Wallet", "cardano-wallet", 100)
+    addrs1 <- listAddresses ctx wDest1
+    addrs2 <- listAddresses ctx wDest2
+    addrs3 <- listAddresses ctx wDest3
+
+    -- each amt is going to different address
+    let addrIds = view #id <$> take n (addrs1 ++ addrs2 ++ addrs3)
+    -- when sending to everything to one address - it worked
+    -- let addrIdRepl = replicate n addrIds !! 0
+    let amounts = take n [1, 1..] :: [Natural]
```
Comment (rvl) [inline lib/core/test/integration/Test/Integration/Framework/DSL.hs]: Would it be better to remove this line and directory write `"quantity": 1,` in the json?

File: lib/core/test/integration/Test/Integration/Framework/DSL.hs [hunk 2]
```diff
@@ -613,6 +615,65 @@ emptyWalletWith ctx (name, passphrase, addrPoolGap) = do
...
+    wSrc <- fixtureWalletWith ctx (replicate n amt)
+    wDest1 <- emptyWalletWith ctx ("Wallet", "cardano-wallet", 100)
+    wDest2 <- emptyWalletWith ctx ("Wallet", "cardano-wallet", 100)
+    wDest3 <- emptyWalletWith ctx ("Wallet", "cardano-wallet", 100)
+    addrs1 <- listAddresses ctx wDest1
+    addrs2 <- listAddresses ctx wDest2
+    addrs3 <- listAddresses ctx wDest3
+
+    -- each amt is going to different address
+    let addrIds = view #id <$> take n (addrs1 ++ addrs2 ++ addrs3)
```
Comment (rvl) [inline lib/core/test/integration/Test/Integration/Framework/DSL.hs]: I don't quite understand. Why are we concatenating the addresses from three different wallets? Is it so this fixture can provide up to 300 addresses?

File: lib/core/test/integration/Test/Integration/Scenario/API/Transactions.hs [hunk 3]
```diff
@@ -655,7 +639,7 @@ spec = do
             expectErrorMessage errMsg405 r
 
 
-    describe "TRANS_ESTIMATE_02 - HTTP headers" $ do
+    describe "TRANS_ESTIMATE_08 - HTTP headers" $ do
```
Comment (rvl) [inline lib/core/test/integration/Test/Integration/Scenario/API/Transactions.hs]: Does this need to be `TRANS_ESTIMATE_09`?

I was thinking about how to reduce test ID churn. What about this technique which was used by BASIC programmers. Number test cases with increments of ten. E.g. TEST_0010, TEST_0020, TEST_0030. Then when you need to insert a test case, you can call it TEST_0015. :nerd_face:

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        (wSrc, payload) <- fixtureZeroAmtSingle ctx
+        r <- request @(ApiTransaction t) ctx (postTxEp wSrc) Default payload
+        expectResponseCode HTTP.status202 r
+
+    it "TRANS_CREATE_09 - 0 amount transaction is accepted on multi-output tx" $ \ctx -> do
+        (wSrc, payload) <- fixtureZeroAmtMulti ctx
+        r <- request @(ApiTransaction t) ctx (postTxEp wSrc) Default payload
+        expectResponseCode HTTP.status202 r
+
+    it "TRANS_ESTIMATE_09 - 0 tx fee estimation is accepted on single output tx" $ \ctx -> do
```
Comment (rvl) [inline lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs]: I don't get this name.

Is it something like "a fee can be estimated for a tx with a single output of amount 0"?

Do we need to assert anything about the fee? e.g. > 0

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        fee <- request @ApiFee ctx (postTxFeeEp wSrc) Default payload
+        tx <- request @(ApiTransaction t) ctx (postTxEp wSrc) Default payload
+        verify fee
+            [ expectResponseCode HTTP.status403
+            , expectErrorMessage "I can't estimate transaction fee because\
+                \ transaction has either more than 255 inputs or more than\
+                \ 255 outputs."
+            ]
+        verify tx
+            [ expectResponseCode HTTP.status500
```
Comment (rvl) [inline lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs]: Should we put a `FIXME` on this line?

**Review by rvl — APPROVED** [review]
Good stuff - a few questions/comments from me.

File: lib/core/test/integration/Test/Integration/Framework/DSL.hs (hunk 1 above)
Comment (piotr-iohk) [reply lib/core/test/integration/Test/Integration/Framework/DSL.hs]: Indeed, that would be better :)

File: lib/core/test/integration/Test/Integration/Framework/DSL.hs (hunk 2 above)
Comment (piotr-iohk) [reply lib/core/test/integration/Test/Integration/Framework/DSL.hs]: > Is it so this fixture can provide up to 300 addresses?

Yes. I noticed it behaves differently when you send transaction with a lot of inputs (256) to different addresses (e.g. got `403` on fee estimation and `500` on post tx) and different when you send it to the same address (both fee estimation and post tx returned `202`). Will look into it now after the fix of #522.

File: lib/core/test/integration/Test/Integration/Scenario/API/Transactions.hs (hunk 3 above)
Comment (piotr-iohk) [reply lib/core/test/integration/Test/Integration/Scenario/API/Transactions.hs]: >Does this need to be TRANS_ESTIMATE_09?

The main reason is that I'm trying to track it with the document:
https://docs.google.com/spreadsheets/d/1fadIA_j4nCd3FbylPo5J8K9Fy6tixSC2T9V3xHKuUvU

The document gathers capabilities/requirements rather from the API/CLI consumer standpoint and helps me track which areas are covered and where are potential gaps.

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs (hunk 4 above)
Comment (piotr-iohk) [reply lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs]: > Is it something like "a fee can be estimated for a tx with a single output of amount 0"?

Yes. This is much better. :)

>Do we need to assert anything about the fee? e.g. > 0

Yes. Good point.

File: lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs (hunk 5 above)
Comment (piotr-iohk) [reply lib/jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/API/Transactions.hs]: Actually this was just to showcase that there is something is wrong, which eventually turned out to be #522. I'll change it. We'll see also if this test will actually be merged as it takes ~5minutes to run, (preparing the 256 utxo takes a lot of time)... I need to play around.

**Comment by piotr-iohk** [comment]
Closing this one because of the mess rebase caused... I'll prepare  another pr with those tests.

**Pull request closed without merging**